cip/1.accepted/CIP2017-02-06-Regular-Path-Patterns.adoc

```diff
@@ -0,0 +1,160 @@
+
+[source, cypher]
+----
+MATCH (a)-//-(b)
+----
+
+This pattern simply states that `a` and `b` must be the same node, and is thus the same as:
+
+[source, cypher]
+----
+MATCH (a), (b) WHERE a = b
+----
+
+The same reader will also have noticed that it is possible to define a pattern containing just a relationship type:
+
+[source, cypher]
+----
+MATCH (a)-/:KNOWS/->(b)
+----
+
+That pattern is indeed equivalent to the very similar relationship pattern:
+
+[source, cypher]
+----
+MATCH (a)-[:KNOWS]->(b)
+----
+
+The main difference being that the variant with a relationship pattern is able to bind that relationship and express further predicates over it.
+
+The Regular Path Patterns start becoming interesting when larger expressions are put together:
+
+[source, cypher]
+.Finding someone loved by someone hated by someone you know, transitively
+----
+MATCH (you)-/(:KNOWS :HATES)+ :LOVES/->(someone)
+----
+
+Note the `+` expressing one or more occurrences of the sequence `KNOWS` followed by `HATES`.
+
+The direction of each relationship is governed by the overall direction of the Regular Path Pattern.
+It is however possible to explicitly define the direction for a particular part of the pattern.
+This is done by either prefixing that part with `<` for a right-to-left direction or suffixing it with `>` for a left-to-right direction.
+It is possible to both prefix the part with `<` and suffixing it with `>`, giving that part the interpretation of being undirected.
+
+[source, cypher]
+.Specifying the direction for different parts of the pattern
+----
+MATCH (you)-/(:KNOWS <:HATES)+ :LOVES/->(someone)
+----
+
+In the example above we say that the `HATES` relationships should have the opposite direction to the other relationships in the path.
+
+Through the use of Defined Path Predicates we can express even more predicates over a path:
+
+[source, cypher]
+.Find a chain of unreciprocated lovers
+----
+MATCH (you)-/unreciprocated_love*/->(someone)
+PATH (a)-/unreciprocated_love/->(b) IS
+     (a)-[:LOVES]->(b)
+     WHERE NOT EXISTS { (b)-[:LOVES]->(a) }
+----
+
+Note how there is no colon used for referencing the Defined Path Predicate, the colon is used in Regular Path Patterns only for referencing actual relationship types.
+
+Sometimes it will be interesting to express a predicate on a node in a Regular Path Pattern.
+This can be achieved by using a Defined Path Predicate where the nodes on both ends are the same:
+
+[source, cypher]
+.Find friends of friends that are not haters
+----
+MATCH (you)-/:KNOWS not_a_hater :KNOWS/-(friendly_friend_of_friend)
+PATH (x)-/not_a_hater/-(x) IS (x)
+     WHERE NOT EXISTS { (x)-[:HATES]->() }
+----
+
+In the case of a Defined Path Predicate where both nodes are the same, the direction of the predicate is irrelevant.
+In general the direction of a Defined Path Predicate is quite important, and used for mapping the pattern in the predicate into the Regular Path Patterns that reference it.
```

Comment:
This applies to the definition of a predicate, right? I imagine that it will be perfectly possible to reverse the direction in the Regular Path Pattern, like so:

```
MATCH (a)<-/pred/-(b)
PATH (a)-/pred/->(b) IS
     (a)-[:KNOWS]->(b)
```

How about it being used in an undirected Regular Path Pattern?

```
MATCH (a)-/pred/-(b)
PATH (a)-/pred/->(b) IS
     (a)-[:KNOWS]->(b)
```